### PR TITLE
 MAILBOX-332 Cassandra implementation for Quota Threshold Crossing notification

### PR DESCRIPTION
**Pull request by chibenwa**

### CODE REVIEW

**Comment by chibenwa** [comment]
test this please

File: mailbox/plugin/quota-mailing-cassandra/src/main/java/org/apache/james/mailbox/quota/cassandra/dto/QuotaThresholdChangedEventDTO.java [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+import org.apache.james.eventsourcing.EventId;
+import org.apache.james.eventsourcing.cassandra.dto.EventDTO;
+import org.apache.james.mailbox.quota.mailing.aggregates.UserQuotaThresholds;
+import org.apache.james.mailbox.quota.mailing.events.QuotaThresholdChangedEvent;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonIgnore;
+import com.fasterxml.jackson.annotation.JsonProperty;
+
+public class QuotaThresholdChangedEventDTO implements EventDTO {
```
Comment (mbaechler) [inline mailbox/plugin/quota-mailing-cassandra/src/main/java/org/apache/james/mailbox/quota/cassandra/dto/QuotaThresholdChangedEventDTO.java]: all DTO should be package-local, don't you think ?

File: mailbox/plugin/quota-mailing-cassandra/src/test/java/org/apache/james/eventsourcing/cassandra/JsonEventSerializerTest.java [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+
+import org.apache.james.eventsourcing.EventId;
+import org.apache.james.eventsourcing.TestAggregateId;
+import org.apache.james.eventsourcing.TestEvent;
+import org.apache.james.eventsourcing.cassandra.dto.TestEventDTOModule;
+import org.junit.jupiter.api.Test;
+
+class JsonEventSerializerTest {
```
Comment (mbaechler) [inline mailbox/plugin/quota-mailing-cassandra/src/test/java/org/apache/james/eventsourcing/cassandra/JsonEventSerializerTest.java]: we need a test that demonstrates ability to be polymorphic (ie deserialize to two different types)

File: mailbox/plugin/quota-mailing-cassandra/src/test/java/org/apache/james/mailbox/quota/cassandra/dto/DTOTest.java [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+    public static final String NO_CHANGES_JSON = "{" +
+        "  \"recentness\":null," +
+        "  \"threshold\":null," +
+        "  \"instant\":null," +
+        "  \"change\":\"NoChange\"" +
+        "}";
+
+    @Test
+    void shouldSerializeQuotaDTO() throws Exception {
+        ObjectMapper objectMapper = new ObjectMapper();
```
Comment (mbaechler) [inline mailbox/plugin/quota-mailing-cassandra/src/test/java/org/apache/james/mailbox/quota/cassandra/dto/DTOTest.java]: why do we keep repeating that 2 lines ? aren't we supposed to use the same as in our production code (with same config) ?

File: mailbox/plugin/quota-mailing-cassandra/src/main/java/org/apache/james/eventsourcing/cassandra/CassandraEventStoreModule.java [hunk 4]
```diff
@@ -39,7 +40,7 @@ public CassandraEventStoreModule() {
                 SchemaBuilder.createTable(CassandraEventStoreTable.EVENTS_TABLE)
                     .ifNotExists()
                     .addPartitionKey(CassandraEventStoreTable.AGGREGATE_ID, DataType.varchar())
-                    .addClusteringColumn(CassandraEventStoreTable.EVENT_ID, DataType.cint())
+                    .addClusteringColumn(CassandraEventStoreTable.EVENT_ID, DataType.bigint())
```
[the remark this answers is not in the document]
Comment (chibenwa) [reply mailbox/plugin/quota-mailing-cassandra/src/main/java/org/apache/james/eventsourcing/cassandra/CassandraEventStoreModule.java]: Event id is a `long` , cassandra complained that it was not the right type.

[the remark this answers is not in the document]
Comment (mbaechler) [reply mailbox/plugin/quota-mailing-cassandra/src/main/java/org/apache/james/eventsourcing/cassandra/CassandraEventStoreModule.java]: Yes I know, and Event id should be a short (:

File: mailbox/plugin/quota-mailing-cassandra/src/main/java/org/apache/james/mailbox/quota/cassandra/dto/HistoryEvolutionDTO.java [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+        return threshold;
+    }
+
+    public Optional<Long> getInstant() {
+        return instant;
+    }
+
+    @JsonIgnore
+    public HistoryEvolution toHistoryEvolution() {
+        Preconditions.checkState(Booleans.countTrue(
```
[the remark this answers is not in the document]
Comment (chibenwa) [reply mailbox/plugin/quota-mailing-cassandra/src/main/java/org/apache/james/mailbox/quota/cassandra/dto/HistoryEvolutionDTO.java]: Well no...

I decided to store the two properties separately (without wrapping them in yet another DTO) so this change do not concern HistoryEvolution builder.

File: mailbox/plugin/quota-mailing-cassandra/src/test/java/org/apache/james/mailbox/quota/cassandra/dto/DTOTest.java (hunk 3 above)
Comment (chibenwa) [reply mailbox/plugin/quota-mailing-cassandra/src/test/java/org/apache/james/mailbox/quota/cassandra/dto/DTOTest.java]: I had many problems while writing the DTOs.

These tests help to know if the problem came from the DTOs or from the JsonDeserializer code.

But we can not directly use the JsonDeserializer as these are not events. Hence this code.

I'd like to keep these tests as they GREATLY helped me: they have value.

File: mailbox/plugin/quota-mailing-cassandra/src/main/java/org/apache/james/eventsourcing/cassandra/CassandraEventStoreModule.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (chibenwa) [reply mailbox/plugin/quota-mailing-cassandra/src/main/java/org/apache/james/eventsourcing/cassandra/CassandraEventStoreModule.java]: As you wish

File: mailbox/plugin/quota-mailing-cassandra/src/test/resources/json/event.json [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+ },
+ "sizeEvolution": {
+   "change": "NoChange",
+   "threshold": null,
+   "instant": null,
+   "recentness": null
+ },
+ "countEvolution": {
+   "change": "NoChange",
+   "threshold": null,
```
Comment (mbaechler) [inline mailbox/plugin/quota-mailing-cassandra/src/test/resources/json/event.json]: we should ignore null fields

File: backends-common/cassandra/pom.xml [hunk 7]
```diff
@@ -30,6 +30,10 @@
     <artifactId>apache-james-backends-cassandra</artifactId>
     <name>Apache James Cassandra backend</name>
 
+    <properties>
+        <cassandra.driver.version>3.5.0</cassandra.driver.version>
```
Comment (mbaechler) [inline backends-common/cassandra/pom.xml]: this should be a separate commit

File: backends-common/cassandra/src/main/java/org/apache/james/backends/cassandra/init/ClusterBuilder.java [hunk 8]
```diff
@@ -178,6 +179,8 @@ public Cluster build() {
         try {
             queryLogger.map(queryLoggerConfiguration ->
                 cluster.register(queryLoggerConfiguration.getQueryLogger()));
+            cluster.getConfiguration().getCodecRegistry()
```
Comment (mbaechler) [inline backends-common/cassandra/src/main/java/org/apache/james/backends/cassandra/init/ClusterBuilder.java]: this is not needed

File: backends-common/cassandra/pom.xml (hunk 7 above)
Comment (mbaechler) [reply backends-common/cassandra/pom.xml]: done

File: backends-common/cassandra/src/main/java/org/apache/james/backends/cassandra/init/ClusterBuilder.java (hunk 8 above)
Comment (mbaechler) [reply backends-common/cassandra/src/main/java/org/apache/james/backends/cassandra/init/ClusterBuilder.java]: done

File: mailbox/plugin/quota-mailing-cassandra/src/test/resources/json/event.json (hunk 6 above)
Comment (mbaechler) [reply mailbox/plugin/quota-mailing-cassandra/src/test/resources/json/event.json]: done

File: mailbox/plugin/quota-mailing-cassandra/src/test/java/org/apache/james/mailbox/quota/cassandra/dto/DTOTest.java (hunk 3 above)
Comment (mbaechler) [reply mailbox/plugin/quota-mailing-cassandra/src/test/java/org/apache/james/mailbox/quota/cassandra/dto/DTOTest.java]: I generally don't agree on such arguments : what have value during dev is not to be kept everytime.
We should keep as little tests as possible that add value to future devs.
As the solution is already in the production code, having some tests explaining how to use jackson API is a bit duplicate.
Finally, I'm not against keeping it, but just be careful about what we expect from unit test.

**Comment by mbaechler** [comment]
I squashed commits and split some things out too.

**Comment by chibenwa** [comment]
:+1:

**Comment by chibenwa** [comment]
Merged
